File: 0149-max-points-on-a-line/0149-max-points-on-a-line.py
class Solution:
  def maxPoints(self, points: List[List[int]]) -> int:
    if len(points) < 3:
      return len(points)
    
    groups = collections.defaultdict(set)
    def group(pointA: List[int], pointB: List[int]) -> None:
      slope = float('inf') if pointB[0] == pointA[0] else (pointB[1]-pointA[1])/(pointB[0]-pointA[0])
      invSlope = float('inf') if pointB[1] == pointA[1] else (pointB[0]-pointA[0])/(pointB[1]-pointA[1])
      yInt = float('inf') if invSlope == 0 else pointA[1]-slope*pointA[0]
      xInt = float('inf') if slope == 0 else pointA[0]-invSlope*pointA[1]
      key = (slope, xInt, yInt)
      groups[key].add(tuple(pointA))
      groups[key].add(tuple(pointB))
      
    for i in range(len(points)-1):
      for j in range(i+1, len(points)):
        group(points[i], points[j])
        
    ans = 0
    for key in groups:
      ans = max(ans, len(groups[key]))
    return ans

File: 0149-max-points-on-a-line/0149-max-points-on-a-line.rb
# @param {Integer[][]} points
# @return {Integer}
def group(point_a, point_b, groups)
  slope = point_b[0] == point_a[0] ? 1/0.0 : 1.0*(point_b[1]-point_a[1])/(point_b[0]-point_a[0])
  inv_slope = point_b[1] == point_a[1] ? 1/0.0 : 1.0*(point_b[0]-point_a[0])/(point_b[1]-point_a[1])
  y_int = inv_slope == 0 ? 1/0.0 : point_a[1]-slope*point_a[0]
  x_int = slope == 0 ? 1/0.0 : point_a[0]-inv_slope*point_a[1]
  key = [slope, x_int, y_int]
  groups[key] << point_a
  groups[key] << point_b
end

def max_points(points)
  return points.length if points.length < 3
  groups = Hash.new{ |h, k| h[k] = Set.new }
  (0...points.length-1).each do |i|
    (i+1...points.length).each do |j|
      group(points[i], points[j], groups)
    end
  end
  
  ans = 0
  groups.each_key{ |key| ans = [ans, groups[key].size].max }
  ans
end